cmd/topology.go

```diff
@@ -58,78 +98,308 @@ func runTopology(cmd *cobra.Command, args []string) error {
+	logf.Log.V(1).Info("Rendered graph to SVG", "buffer length", buf.Len(), "error", err)
 	if err != nil {
 		return err
 	}
 
-	// write to file
-	fSvg, err := os.Create(topologySVGOutputFile)
+	fSvg, err := os.Create(outputFile)
 	if err != nil {
 		return err
 	}
 	defer fSvg.Close()
 
 	_, err = fSvg.Write(buf.Bytes())
-	logf.Log.V(1).Info("write topology in SVG format to file", "file", topologySVGOutputFile, "error", err)
+	logf.Log.V(1).Info("Wrote topology in SVG format to file", "file", outputFile, "error", err)
 	if err != nil {
 		return err
 	}
 
+	return nil
+}
+
+func openSVG(filePath string) error {
 	externalCommand := "xdg-open"
 	if _, err := exec.LookPath("open"); err == nil {
 		externalCommand = "open"
 	}
 
-	openCmd := exec.Command(externalCommand, topologySVGOutputFile)
-	// pipe the commands output to the applications
-	// standard output
+	openCmd := exec.Command(externalCommand, filePath)
 	openCmd.Stdout = os.Stdout
+	openCmd.Stderr = os.Stderr
 
-	// Run still runs the command and waits for completion
-	// but the output is instantly piped to Stdout
-	if err := openCmd.Run(); err != nil {
-		return err
+	return openCmd.Start()
+}
+
+func handleWebSocket(w http.ResponseWriter, r *http.Request, topologyConfigMap *corev1.ConfigMap, mgr manager.Manager, ctx context.Context) {
+	conn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		logf.Log.Error(err, "Failed to upgrade to WebSocket")
+		return
 	}
 
-	return nil
+	connectionsMu.Lock()
+	connections[conn] = true
+	connectionsMu.Unlock()
+
+	logf.Log.Info("New WebSocket connection established")
+
+	err = conn.WriteMessage(websocket.TextMessage, []byte(topologyConfigMap.Data["topology"]))
+	if err != nil {
+		logf.Log.Error(err, "Failed to send initial topology")
+		removeConnection(conn)
+		return
+	}
+
+	go func() {
+		defer removeConnection(conn)
+		for {
+			_, _, err := conn.ReadMessage()
+			if err != nil {
+				if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
+					logf.Log.Info("WebSocket connection closed gracefully")
+				} else {
+					logf.Log.Error(err, "WebSocket read error")
+				}
+				break
+			}
+		}
+	}()
+}
+
+func removeConnection(conn *websocket.Conn) {
+	connectionsMu.Lock()
+	defer connectionsMu.Unlock()
+	if _, exists := connections[conn]; exists {
+		conn.Close()
+		delete(connections, conn)
+		logf.Log.Info("WebSocket connection removed")
+	}
+}
+
+func broadcastTopologyUpdate(topology string) {
+	connectionsMu.Lock()
+	defer connectionsMu.Unlock()
+	for conn := range connections {
+		err := conn.WriteMessage(websocket.TextMessage, []byte(topology))
+		if err != nil {
+			logf.Log.Error(err, "Failed to send topology update to a client")
+			conn.Close()
+			delete(connections, conn)
+		}
+	}
+}
+
+func (w *configMapWatcher) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	configMap := &corev1.ConfigMap{}
+	if err := w.Get(ctx, req.NamespacedName, configMap); err != nil {
```

Comment:
if the topology is not found, that is a known error you can check and exit without any error. Returning an error would make the controller to retry